### PR TITLE
MEX-71 fixes after review

### PR DESCRIPTION
**Pull request by maksenius**
### Description

Please include a summary of the change and what type of change it is (new feature, bug fix, refactoring, documentation).
Please also include relevant motivation and context.
List any dependencies that are required for this change.

Fixes # (issue)

### Quick checks:

- [ ] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [ ] There is no other [pull request](https://github.com/ConduitIO/conduit-connector-snowflake/pulls) for the same update/change.
- [ ] I have written unit tests.
- [ ] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

File: config/config_test.go [hunk 1]
```diff
@@ -32,22 +32,40 @@ func TestParse(t *testing.T) {
 			cfg: map[string]string{
 				"connection": "user:password@my_organization-my_account/mydb/public",
 				"table":      "customer",
-				"key":        "id",
+				"primaryKey": "id",
```
Comment (maha-hajja) [inline config/config_test.go:35]: we have all these key values as constants in config.go, I think we should use them instead, for all the tests

Comment (maksenius) [reply config/config_test.go]: Done

File: repository/snowflake.go [hunk 2]
```diff
@@ -230,7 +230,7 @@ func (s *Snowflake) GetTrackingData(
 func buildGetDataQuery(table, key string, fields []string, offset, limit int) string {
 	sb := sqlbuilder.NewSelectBuilder()
 
-	if fields == nil {
+	if fields == nil || len(fields) == 0 {
```
Comment (hariso) [inline repository/snowflake.go:233]: len(fields) should work for nil slices too, so I believe you can have only len(fields) == 0

File: source/position/position.go [hunk 3]
```diff
@@ -34,15 +34,15 @@ type Position struct {
...
 	IteratorType IteratorType
 
-	// Element - index position in current offset.
-	Element int
-	// Offset - show current offset.
-	Offset int
+	// IndexInBatch - index position in current batch.
+	IndexInBatch int
+	// BatchID - batch id.
+	BatchID int
```
Comment (hariso) [inline source/position/position.go:40]: Now I remembered one thing... Connectors can be reconfigured (through the UI or the API), which means that the batch size can change. And if the batch size changes, the same position can mean a completely different record, right?

File: repository/snowflake.go (hunk 2 above)
Comment (maksenius) [reply repository/snowflake.go]: Fixed

File: source/position/position.go (hunk 3 above)
Comment (hariso) [reply source/position/position.go:40]: IMHO, it's better to have it configurable (when creating the connector), than to have it modifiable.

File: README.md [hunk 4]
```diff
@@ -34,37 +35,74 @@ Run `make test`.
...
 
 Position has fields: `type` (`c` - CDC or `s`- Snapshot), `element`(index of element of current
-offset), `offset`.
+offset), `offset`. <br> 
+Example: 
+```
+{
+ "Type" : "c",
+ "Element": 2,
+ "Offset": 5
```
Comment (hariso) [inline README.md:106]: These were renamed to BatchID and IndexInBatch.

File: README.md [hunk 5]
```diff
@@ -34,37 +35,74 @@ Run `make test`.
...
 #### Snapshot Iterator
  
-The snapshot iterator starts get data from table using select query with limit and offset. Batch size is 1000,
-offset value is zero for first time. Iterator save information from table to `data` slice variable.
-Iterator `HasNext` method check if next element exist in `data` using variable `index` and if it is needed
-change offset and run select query to get new data with new offset. Method `Next` gets next element and converts 
-it to `Record` sets metadata variable `table`, set metadata variable `action` - `insert`, increase `index`.
+The snapshot iterator starts getting data from the table using select query with limit and offset. 
+Batch size is configurable, offset value is zero for first time. Iterator save information from table 
+to `data` slice variable. <br>
```
Comment (hariso) [inline README.md:48]: > Iterator save information from table  to `data` slice variable. <br>

IIUC, we can also say "The current batch of rows is saved in the `data` slice in the iterator." in which case we can maybe rename `data` -> `currentBatch` or something like that?

nitpick: I think you can use two newlines instead of `<br>`

File: README.md [hunk 6]
```diff
@@ -34,37 +35,74 @@ Run `make test`.
...
+Iterator `HasNext` method check if next element exist in `data` using variable `index`
+and if it is needed change offset and run select query to get new data with new offset. Method `Next` gets 
+next element and converts it to `Record` sets metadata variable `table`, set metadata variable `action` - `insert`,
+increase `index`.<br>
+If snapshot stops, it will parse position from last record. Position has fields: `Element` - it is the index of element 
+in current batch, this last element what was recorded, `Offset` - shows the last value offset what iterator used in get data
+query. Iterator runs query to get data from table with `batchSize` and `offset` which was got from position. 
+`index` value will be `Element` increased by one, because iterator tries to find next element in current batch.
+If `index` > `batchSize` iterator will change `offset` to next and set `index` zero.
```
Comment (hariso) [inline README.md:57]: So, through an example: offset = 10, element = 1, batch size = 20. That means that the last processed row was 20*10+1 = 201, correct? If the snapshot is stopped here, and then started it will do something like `select * from mytable offset 201`?

File: README.md [hunk 7]
```diff
@@ -34,37 +35,74 @@ Run `make test`.
...
+
+`Stream` has columns:<br>
+`METADATA$ACTION`: Indicates the DML operation (INSERT, DELETE) recorded.<br>
+`METADATA$ISUPDATE`: Indicates whether the operation was part of an UPDATE statement.
+Updates to rows in the source object are represented as a pair of DELETE and
+INSERT records in the stream with a metadata column METADATA$ISUPDATE values set to TRUE.<br>
+`METADATA$ROW_ID`: Specifies the unique and immutable ID for the row, which can be used to track changes 
+to specific rows over time. <br>
+
 When source starts work first time iterator <b>creates</b> stream with name `conduit_stream_{table}` to `table` from
```
[the remark this answers is not in the document]
Comment (maksenius) [reply README.md:81]: Yeah, you are right

File: README.md (hunk 6 above)
Comment (maksenius) [reply README.md:57]: For example we got last position and stopped source. Batch size is 10.
```{
"IteratorType" : "S",
"IndexInBatch": 1,
"BatchID" : 20,
}```

When we started connector next time  we run query
``` select * from mytable offset 20 limit 10;```

Last processed element was 21 . Offset in snowflake is the row number after which the limited rows are returned.
It is means , when we processed all elements in this batch, next batchID will be 30

File: source/position/position.go (hunk 3 above)
Comment (maksenius) [reply source/position/position.go:40]: I think we can remove batchID from position, and leave field `Index` - number of row in table.  if source stopped we can continue get data  from using query where index will be offset. What do you thing about it?

File: README.md (hunk 4 above)
Comment (maksenius) [reply README.md]: fixed

File: README.md (hunk 5 above)
Comment (maksenius) [reply README.md]: renamed

File: acceptance_test.go [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+				NewSpecification: Specification,
+				NewSource:        source.New,
+				NewDestination:   nil,
+			},
+			SourceConfig:      cfg,
+			DestinationConfig: nil,
+			GoleakOptions: []goleak.Option{
+				goleak.IgnoreTopFunction("database/sql.(*DB).connectionOpener"),
+				goleak.IgnoreTopFunction("net/http.(*persistConn).writeLoop"),
+				goleak.IgnoreTopFunction("internal/poll.runtime_pollWait"),
```
Comment (hariso) [inline acceptance_test.go:120]: It would be good to leave a comment saying why do we have these leaks and why we ignore them (impossible to fix, false positives etc.)

File: README.md [hunk 9]
```diff
@@ -34,37 +34,127 @@ Run `make test`.
...
+Iterator `HasNext` method check if next element exist in `currentBatch` using variable `index`
+and if it is needed change offset and run select query to get new data with new offset. Method `Next` gets
+next element and converts it to `Record` sets metadata variable `table`, set metadata variable `action` - `insert`,
+increase `index`.
+
+If snapshot stops, it will parse position from last record. Position has fields: `Element` - it is the index of element
+in current batch, this last element what was recorded, `Offset` - shows the last value offset what iterator used for
+getting data query. Iterator runs query to get data from table with `batchSize` and `offset` which was got from
+position. `index` value will be `Element` increased by one, because iterator tries to find next element in current batch.
+If `index` > `batchSize` iterator will change `offset` to next and set `index` zero.
```
Comment (hariso) [inline README.md:58]: Looks good to me. I have only one tiny comment, which is to replace all `Element` and `Offset` usages to `IndexInBatch` and `BatchId`.

File: docs/CDC Iterator.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
+# CDC Iterator implementation, issues, ideas.
+
+## General.
+
+Snowflake doesn't have ability to provide information about data changes without stream creation.
+Team have some ideas how to design it. Some of these ideas in this document.
```
Comment (hariso) [inline docs/CDC Iterator.md:6]: We can maybe rephrase as: "A few options for supporting CDC in Snowflake have been discussed and they are mentioned in this document". It would also be good to document all the ideas and their pros and cons, instead of just some of them.

File: acceptance_test.go (hunk 8 above)
Comment (maksenius) [reply acceptance_test.go]: Added. 2 leaks is problem with snowflake driver, I decided to create issue

File: README.md [hunk 11]
```diff
@@ -34,37 +34,127 @@ Run `make test`.
...
+`METADATA$ISUPDATE`: Indicates whether the operation was part of an UPDATE statement.
+Updates to rows in the source object are represented as a pair of DELETE and
+INSERT records in the stream with a metadata column METADATA$ISUPDATE values set to TRUE.<br>
+`METADATA$ROW_ID`: Specifies the unique and immutable ID for the row, which can be used to track changes
+to specific rows over time.
+
 When source starts work first time iterator <b>creates</b> stream with name `conduit_stream_{table}` to `table` from
-config, <b>creates</b> table for consuming stream with name `conduit_tracking_{table}`. 
+config, <b>creates</b> table for consuming stream with name `conduit_tracking_{table}`.
 This consuming table has the same schema as `table`  with additional metadata columns:
```
[the remark this answers is not in the document]
Comment (maksenius) [reply README.md]: rephrased and extended it

File: README.md [hunk 12]
```diff
@@ -20,8 +20,8 @@ The config passed to `Configure` can contain the following fields.
...
-| `key`        | Column name that records should use for their `Key` fields.                                                                                                                                                                                     | yes      | "id"                                                   |
-
+| `primaryKey` | Column name that records should use for their `Key` fields.                                                                                                                                                                                     | yes      | "id"                                                   |
+| `batchSize`  | Size of batch. By default is 1000. <b>Important:</b> Please don't update this variable, it will cause problem with position.                                                                                                                    | no       | "100"                                                  |
```
[the remark this answers is not in the document]
Comment (maksenius) [reply README.md]: Sure, done
